node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1084,75 +1098,191 @@ async fn fetch_to_genesis(trusted_block: &Block, ctx: &ChainSyncContext<'_>) ->
-                // reported lowest available block is actually unavailable. In this case the
-                // available range at startup will be [0, 0]. For all other cases, this is an error.
+                lowest_trusted_block_header = new_lowest;
+            }
+            Err(err) => {
+                // If we get an error here it means something must have gone really wrong.
+                // We either get the data from storage or from a peer where we retry ad infinitum if
+                // peer times out or item is absent. The only reason we would end up
+                // here is if fetcher couldn't construct a fetch request.
                 error!(
-                    locally_available_block_range_on_start =
-                        %ctx.locally_available_block_range_on_start,
-                    "failed to get lowest block reported as available"
+                    ?err,
+                    "failed to download block headers batch with infinite retries"
                 );
+                return Err(err.into());
             }
         }
 
-        let walkback_block_height = walkback_block.height();
-        ctx.metrics
-            .chain_sync_block_height_synced
-            .set(walkback_block_height as i64);
-        info!(%walkback_block_height, "syncing block height");
-        sync_trie_store(*walkback_block.header().state_root_hash(), ctx).await?;
-        ctx.effect_builder
-            .update_lowest_available_block_height_in_storage(walkback_block.height())
-            .await;
-        if walkback_block.height() == 0 {
+        if lowest_trusted_block_header.height() == 0 {
             break;
-        } else {
-            walkback_block = fetch_and_store_block_with_deploys_by_hash(
-                *walkback_block.header().parent_hash(),
-                ctx,
-            )
-            .await?
-            .block
         }
     }
     Ok(())
 }
 
+// Fetches a batch of block headers, validates and stores in storage.
+// Returns either an error or lowest valid block in the chain.
+async fn fetch_block_headers_batch(
+    lowest_trusted_block_header: &BlockHeader,
+    ctx: &ChainSyncContext<'_>,
+) -> Result<BlockHeader, FetchBlockHeadersBatchError> {
+    let batch_id =
+        BlockHeadersBatchId::from_known(lowest_trusted_block_header, MAX_HEADERS_BATCH_SIZE);
+
+    loop {
+        let fetched_headers_data: FetchedData<BlockHeadersBatch> =
+            fetch_retry_forever::<BlockHeadersBatch>(ctx, batch_id).await?;
+        match fetched_headers_data {
+            FetchedData::FromStorage { item } => {
+                return BlockHeadersBatch::validate(
+                    &*item,
+                    &batch_id,
+                    lowest_trusted_block_header,
+                    ctx.config.verifiable_chunked_hash_activation(),
+                )
+                .map_err(FetchBlockHeadersBatchError::InvalidBatchFromStorage)
```

Comment:
Seems like we're calling `validate` here only to get the lowest block header from the batch.  I'd be inclined to change the validation to return `Result<(), ...>` and have a separate getter on the `BlockHeadersBatch` for the lowest header.

Also, nitpick: this would more normally be written `item.validate(...)` rather than `BlockHeadersBatch::validate(&*item, ...)`.

Reply:
> Also, nitpick: this would more normally be written item.validate(...) rather than BlockHeadersBatch::validate(&*item, ...).

I agree but we have the same method name (`validate`) on `Item` trait and it was resolving to that one instead of `BlockHeadersBatch`.

Reply:
>Seems like we're calling validate here only to get the lowest block header from the batch. 

Not really, I also wanted to validate the batch from the storage. Although, if we agree that we should trust data from the DB I will remove that and simply get the lowest element.

Reply:
Done in [`f46bbd2` (#3020)](https://github.com/casper-network/casper-node/pull/3020/commits/f46bbd27aa8bf8d72a06a0b483d09e29ddc84132)